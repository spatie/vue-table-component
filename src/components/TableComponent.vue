<template>
    <div class="table-component">
        <div v-if="showFilter && filterableColumnExists" class="table-component__filter">
            <input
                    :class="fullFilterInputClass"
                    type="text"
                    v-model="filter"
                    :placeholder="filterPlaceholder"
            >
            <a
                    v-if="filter"
                    @click="filter = ''"
                    class="table-component__filter__clear"
            >Ã—</a>
        </div>

        <div class="table-component__table-wrapper">
            <table :class="fullTableClass">
                <caption v-if="showCaption" class="table-component__table__caption" role="alert" aria-live="polite">
                    {{ ariaCaption }}
                </caption>
                <thead :class="fullTableHeadClass">
                <tr>
                    <table-column-header
                            @click="changeSorting"
                            v-for="column in columns"
                            :key="column.show"
                            :sort="sort"
                            :column="column"
                    ></table-column-header>
                </tr>
                </thead>
                <tbody :class="fullTableBodyClass">
                <table-row
                        v-for="row in displayedRows"
                        :key="row.vueTableComponentInternalRowId"
                        :row="row"
                        :columns="columns"
						:class="fullTableRowClass(row)"
						@rowClick="emitRowClick"
                ></table-row>
                </tbody>
                <tfoot>
                    <slot name="tfoot" :rows="rows"></slot>
                </tfoot>
            </table>
        </div>

        <div v-if="displayedRows.length === 0" class="table-component__message">
            {{ filterNoResults }}
        </div>


        <div style="display:none;">
            <slot></slot>
        </div>

        <pagination v-if="pagination" :pagination="pagination" @pageChange="pageChange"></pagination>
    </div>
</template>

<script>
    import Column from '../classes/Column';
    import expiringStorage from '../expiring-storage';
    import Row from '../classes/Row';
    import TableColumnHeader from './TableColumnHeader';
    import TableRow from './TableRow';
    import settings from '../settings';
    import Pagination from './Pagination';
    import { classList, pick } from '../helpers';

    export default {
        components: {
            TableColumnHeader,
            TableRow,
            Pagination,
        },

        props: {
            data: { default: () => [], type: [Array, Function] },

            showFilter: { default: true },
            showCaption: { default: true },

            sortBy: { default: '', type: String },
            sortOrder: { default: '', type: String },

            cacheKey: { default: null },
            cacheLifetime: { default: 5 },

            tableClass: { default: settings.tableClass },
            theadClass: { default: settings.theadClass },
			tbodyClass: { default: settings.tbodyClass },
			rowClass:  { default: settings.rowClass },
            filterInputClass: { default: settings.filterInputClass },
            filterPlaceholder: { default: settings.filterPlaceholder },
            filterNoResults: { default: settings.filterNoResults },
        },

        data: () => ({
            columns: [],
            rows: [],
            filter: '',
            sort: {
                fieldName: '',
                order: '',
            },
            pagination: null,

            localSettings: {},
        }),

        created() {
            this.sort.fieldName = this.sortBy;
            this.sort.order = this.sortOrder;

            this.restoreState();
        },

        async mounted() {
            const columnComponents = this.$slots.default
                .filter(column => column.componentInstance)
                .map(column => column.componentInstance);

            this.columns = columnComponents.map(
                column => new Column(column)
            );

            columnComponents.forEach(columnCom => {
                Object.keys(columnCom.$options.props).forEach(
                    prop => columnCom.$watch(prop, () => {
                        this.columns = columnComponents.map(
                            column => new Column(column)
                        );
                    })
                );
            });

            await this.mapDataToRows();
        },

        watch: {
            filter() {
                if (!this.usesLocalData) {
                    this.mapDataToRows();
                }

                this.saveState();
            },

            data() {
                if (this.usesLocalData) {
                    this.mapDataToRows();
                }
            },
        },

        computed: {
            fullTableClass() {
                return classList('table-component__table', this.tableClass);
            },

            fullTableHeadClass() {
                return classList('table-component__table__head', this.theadClass);
            },

            fullTableBodyClass() {
                return classList('table-component__table__body', this.tbodyClass);
            },

            fullFilterInputClass() {
                return classList('table-component__filter__field', this.filterInputClass);
            },

            ariaCaption() {
                if (this.sort.fieldName === '') {
                    return 'Table not sorted';
                }

                return `Table sorted by ${this.sort.fieldName} ` +
                    (this.sort.order === 'asc' ? '(ascending)' : '(descending)');
            },

            usesLocalData() {
                return Array.isArray(this.data);
            },

            displayedRows() {
                if (!this.usesLocalData) {
                    return this.sortedRows;
                }

                if (!this.showFilter) {
                    return this.sortedRows;
                }

                if (!this.columns.filter(column => column.isFilterable()).length) {
                    return this.sortedRows;
                }

                return this.sortedRows.filter(row => row.passesFilter(this.filter));
            },

            sortedRows() {
                if (!this.usesLocalData) {
                    return this.rows;
                }

                if (this.sort.fieldName === '') {
                    return this.rows;
                }

                if (this.columns.length === 0) {
                    return this.rows;
                }

                const sortColumn = this.getColumn(this.sort.fieldName);

                if (!sortColumn) {
                    return this.rows;
                }

                return this.rows.sort(sortColumn.getSortPredicate(this.sort.order, this.columns));
            },

            filterableColumnExists() {
                return this.columns.filter(c => c.isFilterable()).length > 0;
            },

            storageKey() {
                return this.cacheKey
                    ? `vue-table-component.${this.cacheKey}`
                    : `vue-table-component.${window.location.host}${window.location.pathname}${this.cacheKey}`;
            },
        },

        methods: {
            async pageChange(page) {
                this.pagination.currentPage = page;

                await this.mapDataToRows();
            },

            async mapDataToRows() {
                const data = this.usesLocalData
                    ? this.prepareLocalData()
                    : await this.fetchServerData();

                let rowId = 0;

                this.rows = data
                    .map(rowData => {
                        rowData.vueTableComponentInternalRowId = rowId++;
                        return rowData;
                    })
                    .map(rowData => new Row(rowData, this.columns));
            },

            prepareLocalData() {
                this.pagination = null;

                return this.data;
            },

            async fetchServerData() {
                const page = this.pagination && this.pagination.currentPage || 1;

                const response = await this.data({
                    filter: this.filter,
                    sort: this.sort,
                    page: page,
                });

                this.pagination = response.pagination;

                return response.data;
            },

            async refresh() {
                await this.mapDataToRows();
            },

            changeSorting(column) {
                if (this.sort.fieldName !== column.show) {
                    this.sort.fieldName = column.show;
                    this.sort.order = 'asc';
                } else {
                    this.sort.order = (this.sort.order === 'asc' ? 'desc' : 'asc');
                }

                if (!this.usesLocalData) {
                    this.mapDataToRows();
                }

                this.saveState();
            },

            getColumn(columnName) {
                return this.columns.find(column => column.show === columnName);
            },

            saveState() {
                expiringStorage.set(this.storageKey, pick(this.$data, ['filter', 'sort']), this.cacheLifetime);
            },

            restoreState() {
                const previousState = expiringStorage.get(this.storageKey);

                if (previousState === null) {
                    return;
                }

                this.sort = previousState.sort;
                this.filter = previousState.filter;

                this.saveState();
            },

			emitRowClick(row) {
				this.$emit('rowClick', row);
			},

            fullTableRowClass(row) {
				const classResult = typeof this.rowClass === 'function' ? `${this.rowClass(row)}` : this.rowClass
                return classList('table-component__table__body__row', classResult);
            },
		},
    };
</script>
