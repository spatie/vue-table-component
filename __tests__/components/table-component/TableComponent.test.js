import TableComponent from '../../../src';
import Vue from 'vue/dist/vue.js';
import LocalStorageMock from '../../helpers/LocalStorageMock';
import simulant from 'simulant';

const localStorage = new LocalStorageMock();

window.localStorage = localStorage;

describe('TableComponent', () => {
    Vue.use(TableComponent);

    beforeEach(() => {
        localStorage.clear();
    });

    it('can mount', async () => {
        document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="[{ firstName: 'John' },{ firstName: 'Paul' }]">
                    <table-column show="firstName" label="First name"></table-column>
                </table-component>
            </div>
        `;

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('has an prop to disable the filter', async () => {
        document.body.innerHTML = `
            <div id="app">
                <table-component :show-filter="false"
                    :data="[{ firstName: 'John' },{ id: 2, firstName: 'Paul' }]">
                    <table-column show="firstName" label="First name"></table-column>
                </table-component>
            </div>
        `;

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('has an prop to disable the caption', async () => {
        document.body.innerHTML = `
            <div id="app">
                <table-component :show-caption="false"
                    :data="[{ firstName: 'John' },{ id: 2, firstName: 'Paul' }]">
                    <table-column show="firstName" label="First name"></table-column>
                </table-component>
            </div>
        `;

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will use the property name as a column heading if label is not set', async () => {
        document.body.innerHTML = `
            <div id="app">
                <table-component :show-filter="false"
                    :data="[{ firstName: 'John' },{ id: 2, firstName: 'Paul' }]">
                    <table-column show="firstName"></table-column>
                </table-component>
            </div>
        `;

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('won\'t use the property name as a column heading if label is an empty string', async () => {
        document.body.innerHTML = `
            <div id="app">
                <table-component :show-filter="false"
                    :data="[{ firstName: 'John' },{ id: 2, firstName: 'Paul' }]">
                    <table-column show="firstName" label=""></table-column>
                </table-component>
            </div>
        `;

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can display a custom message when filtering results in no results', async () => {
        document.body.innerHTML = `
            <div id="app">
                <table-component
                    :extra-settings="{ labels: { filterResultEmpty: 'game over man, game over' } }"
                    :data="[{ firstName: 'John' },{ id: 2, firstName: 'Paul' }]">
                    <table-column show="firstName" label="First name"></table-column>
                </table-component>
            </div>
        `;

        const table = await createVm();

        table.filter = 'this returns nothing';

        await Vue.nextTick(() => {});

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can display a custom placeholder in the filter field', async () => {
        document.body.innerHTML = `
            <div id="app">
                <table-component
                    :extra-settings="{ labels: { filterPlaceholder: 'custom placeholder' } }"
                    :data="[{ firstName: 'John' },{ id: 2, firstName: 'Paul' }]">
                    <table-column show="firstName" label="First name"></table-column>
                </table-component>
            </div>
        `;

        const table = await createVm();

        table.filter = 'this returns nothing';

        await Vue.nextTick(() => {});

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can accept a function to fetch the data', async () => {
        const serverResponse = () => {
            return {
                data: [{ firstName: 'John' },{ id: 2, firstName: 'Paul' }],
            };
        };

        document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="${serverResponse}">
                    <table-column show="firstName" label="First name"></table-column>
                </table-component>
            </div>
        `;

        await createVm();

        await Vue.nextTick(() => {});

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can render pagination when the server responds with pagination data', async () => {
        const serverResponse = () => {
            return {
                data: [{ firstName: 'John' },{ id: 2, firstName: 'Paul' }],
                pagination: {
                    totalPages: 4,
                    currentPage: 2,
                },
            };
        };

        document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="${serverResponse}">
                    <table-column show="firstName" label="First name"></table-column>
                </table-component>
            </div>
        `;

        await createVm();

        await Vue.nextTick(() => {});

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('clicking a link in the pagination will rerender the table', async () => {
        const serverResponse = ({ page }) => {
            return {
                data: [{ firstName: `John ${page}` },{ id: 2, firstName: `Paul ${page}` }],
                pagination: {
                    totalPages: 4,
                    currentPage: page,
                },
            };
        };

        document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="${serverResponse}">
                    <table-column show="firstName" label="First name"></table-column>
                </table-component>
            </div>
        `;

        await createVm();

        await Vue.nextTick(() => {});

        expect(document.body.innerHTML).toMatchSnapshot();

        const thirdPageLink = document.getElementsByClassName('page-link')[2];

        await simulant.fire(thirdPageLink, 'click');

        await Vue.nextTick();
        await Vue.nextTick();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can render a set of footer rows', async () => {
        document.body.innerHTML = `
            <div id="app">
                <table-component :show-caption="false"
                    :data="[
                        { firstName: 'John', numSkills: 2 },
                        { id: 2, firstName: 'Paul', numSkills: 3 },
                        { firstName: 'Total Skills', numSkills: 5, isFooterRow: true }]"
                >
                    <table-column show="firstName" label="First name"></table-column>
                    <table-column show="numSkills" label="Number of skills" data-type="numeric"></table-column>
                </table-component>
            </div>
        `;

        const table = await createVm();

        const footer = table.$el.getElementsByTagName('tfoot')[0];

        expect(footer).toBeTruthy();
        expect(footer.tagName).toEqual('TFOOT');
        expect(footer.children[0].tagName).toEqual('TR');

        expect(document.body.innerHTML).toMatchSnapshot();
    });
});

async function createVm() {
    const vm = new Vue({
        el: '#app',
    });

    await Vue.nextTick(() => {});

    const table = vm.$children[0];

    return table;
}
