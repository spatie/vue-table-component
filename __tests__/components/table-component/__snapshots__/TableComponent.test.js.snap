// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableComponent accepts a function to format values 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder="Filter table…"
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Formatted: John
            </td>
          </tr>
          <tr>
            <td>
              Formatted: Paul
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <div>
      </div>
    </div>
    <!---->
  </div>
</div>

`;

exports[`TableComponent can accept a function to fetch the data 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder="Filter table…"
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <div>
      </div>
    </div>
    <!---->
  </div>
</div>

`;

exports[`TableComponent can display a custom message when filtering results in no results 1`] = `

<div id="app">
  <div class="table-component"
       extra-settings="[object Object]"
  >
    <div class="table-component__filter">
      <input type="text"
             placeholder="Filter table…"
             class="table-component__filter__field"
      >
      <a class="table-component__filter__clear">
        ×
      </a>
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody>
        </tbody>
      </table>
    </div>
    <div class="table-component__message">
      There are no matching rows
    </div>
    <div style="display: none;">
      <div>
      </div>
    </div>
    <!---->
  </div>
</div>

`;

exports[`TableComponent can display a custom placeholder in the filter field 1`] = `

<div id="app">
  <div class="table-component"
       extra-settings="[object Object]"
  >
    <div class="table-component__filter">
      <input type="text"
             placeholder="Filter table…"
             class="table-component__filter__field"
      >
      <a class="table-component__filter__clear">
        ×
      </a>
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody>
        </tbody>
      </table>
    </div>
    <div class="table-component__message">
      There are no matching rows
    </div>
    <div style="display: none;">
      <div>
      </div>
    </div>
    <!---->
  </div>
</div>

`;

exports[`TableComponent can display nested properties 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder="Filter table…"
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <div>
      </div>
    </div>
    <!---->
  </div>
</div>

`;

exports[`TableComponent can mount 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder="Filter table…"
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <div>
      </div>
    </div>
    <!---->
  </div>
</div>

`;

exports[`TableComponent can render pagination when the server responds with pagination data 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder="Filter table…"
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <div>
      </div>
    </div>
    <nav>
      <ul class="pagination justify-content-center">
        <li class="page-item">
          <a class="page-link">
            1
          </a>
        </li>
        <li class="page-item active">
          <a class="page-link">
            2
          </a>
        </li>
        <li class="page-item">
          <a class="page-link">
            3
          </a>
        </li>
        <li class="page-item">
          <a class="page-link">
            4
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>

`;

exports[`TableComponent clicking a link in the pagination will rerender the table 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder="Filter table…"
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              John 1
            </td>
          </tr>
          <tr>
            <td>
              Paul 1
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <div>
      </div>
    </div>
    <nav>
      <ul class="pagination justify-content-center">
        <li class="page-item active">
          <a class="page-link">
            1
          </a>
        </li>
        <li class="page-item">
          <a class="page-link">
            2
          </a>
        </li>
        <li class="page-item">
          <a class="page-link">
            3
          </a>
        </li>
        <li class="page-item">
          <a class="page-link">
            4
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>

`;

exports[`TableComponent clicking a link in the pagination will rerender the table 2`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder="Filter table…"
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              John 3
            </td>
          </tr>
          <tr>
            <td>
              Paul 3
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <div>
      </div>
    </div>
    <nav>
      <ul class="pagination justify-content-center">
        <li class="page-item">
          <a class="page-link">
            1
          </a>
        </li>
        <li class="page-item">
          <a class="page-link">
            2
          </a>
        </li>
        <li class="page-item active">
          <a class="page-link">
            3
          </a>
        </li>
        <li class="page-item">
          <a class="page-link">
            4
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>

`;

exports[`TableComponent has an prop to disable the caption 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder="Filter table…"
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <!---->
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <div>
      </div>
    </div>
    <!---->
  </div>
</div>

`;

exports[`TableComponent has an prop to disable the filter 1`] = `

<div id="app">
  <div class="table-component">
    <!---->
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <div>
      </div>
    </div>
    <!---->
  </div>
</div>

`;

exports[`TableComponent supports a scoped slot inside the table column 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder="Filter table…"
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              John slot
            </td>
          </tr>
          <tr>
            <td>
              Paul slot
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <div>
        slot
      </div>
    </div>
    <!---->
  </div>
</div>

`;

exports[`TableComponent will use the property name as a column heading if label is not set 1`] = `

<div id="app">
  <div class="table-component">
    <!---->
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
              firstName
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <div>
      </div>
    </div>
    <!---->
  </div>
</div>

`;

exports[`TableComponent won't use the property name as a column heading if label is an empty string 1`] = `

<div id="app">
  <div class="table-component">
    <!---->
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead>
          <tr>
            <th role="columnheader"
                scope="col"
                aria-sort="none"
                class="table-component__th--sort"
            >
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <div>
      </div>
    </div>
    <!---->
  </div>
</div>

`;
